### PR TITLE
added caching to wp_old_slug_redirect and tests

### PR DESCRIPTION
**Pull request by pbearne**
Not working yet
the new tests are not work

Trac ticket: https://core.trac.wordpress.org/ticket/36723
---
**This Pull Request is for code review only. Please keep all other discussions in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/query.php [hunk 1]
```diff
@@ -1117,7 +1118,20 @@ function _find_post_by_old_slug( $post_type ) {
 		$query .= $wpdb->prepare( ' AND DAYOFMONTH(post_date) = %d', get_query_var( 'day' ) );
 	}
 
-	$id = (int) $wpdb->get_var( $query );
+	$key          = md5( $query );
+	$last_changed = wp_cache_get( 'last_changed', 'posts' );
+	if ( ! $last_changed ) {
+		$last_changed = microtime();
+		wp_cache_set( 'last_changed', $last_changed, 'posts' );
+	}
```
Comment (peterwilsoncc) [inline src/wp-includes/query.php]:
```suggestion
	$last_changed = wp_cache_get_last_changed( 'posts' );
```

:1234: Helper function can be used throughout.

File: src/wp-includes/query.php [hunk 2]
```diff
@@ -1117,7 +1118,20 @@ function _find_post_by_old_slug( $post_type ) {
...
+	$cache_key = "_find_post_by_old_slug:$key:$last_changed";
+	$cache     = wp_cache_get( $cache_key, 'posts' );
+	if ( $cache !== false ) {
+		$id = $cache;
+	} else {
+		$id = (int) $wpdb->get_var( $query );
+		wp_cache_set( $cache_key, $id, 'posts' );
+	}
```
Comment (peterwilsoncc) [inline src/wp-includes/query.php]:
```suggestion
	if ( false !== $cache ) {
		return $cache;
	}

	$id = (int) $wpdb->get_var( $query );
	wp_cache_set( $cache_key, $id, 'posts' );
```

I find fewer indents make the code easier to read but take this as a nitpick.

File: tests/phpunit/tests/rewrite/oldDateRedirect.php [hunk 3]
```diff
@@ -66,6 +67,36 @@ public function test_old_date_redirect() {
...
+		wp_old_slug_redirect();
+		$this->assertSame( $permalink, $this->old_date_redirect_url );
+
+		$query_count = $wpdb->num_queries;
+
+		$this->go_to( $old_permalink );
+		wp_old_slug_redirect();
+		$this->assertSame( $permalink, $this->old_date_redirect_url );
+
+		$this->assertSame( $query_count, $wpdb->num_queries );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/rewrite/oldDateRedirect.php]: Check if these are failing due to the `go_to()` -- if it uses `WP_Query` then an additional query is expected.

File: tests/phpunit/tests/rewrite/oldDateRedirect.php [hunk 4]
```diff
@@ -66,6 +67,36 @@ public function test_old_date_redirect() {
...
+			)
+		);
+
+		$permalink = user_trailingslashit( get_permalink( self::$post_id ) );
+
+		$this->go_to( $old_permalink );
+		wp_old_slug_redirect();
+		$this->assertSame( $permalink, $this->old_date_redirect_url );
+
+		$query_count = $wpdb->num_queries;
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/rewrite/oldDateRedirect.php]: As noted in the other PR, you can use `get_num_queries()` to avoid messing around with globals.

**Review by peterwilsoncc — COMMENTED** [review]
Added some notes inline.

If the test's `go_to()` method uses `WP_Query` then the number of queries may be increasing as the first query in Query isn't cached.

File: src/wp-includes/query.php [hunk 5]
```diff
@@ -1117,15 +1105,40 @@ function _find_post_by_old_slug( $post_type ) {
...
+	 */
+	$id = apply_filters( 'old_slug_redirect_post_id', $id );
+
+	if ( ! $id ) {
+		return;
+	}
+
+	$link = get_permalink( $id );
+
+	wp_cache_set( $cache_key, $link, 'posts' );
```
Comment (spacedmonkey) [inline src/wp-includes/query.php]:
```suggestion
	wp_cache_set( $cache_key, $id, 'posts' );
```

We should not be caching the link. This value change and is filtered. Let's cache the id instead, it is a smaller cache value.

File: tests/phpunit/tests/rewrite/oldDateRedirect.php [hunk 6]
```diff
@@ -48,6 +48,7 @@ public function tear_down() {
 	}
 
 	public function test_old_date_redirect() {
+		global $wpdb;
```
Comment (spacedmonkey) [inline tests/phpunit/tests/rewrite/oldDateRedirect.php:65]: Why does this need to be added?

File: tests/phpunit/tests/rewrite/oldDateRedirect.php [hunk 7]
```diff
@@ -66,6 +67,37 @@ public function test_old_date_redirect() {
 		$this->assertSame( $permalink, $this->old_date_redirect_url );
 	}
 
+	public function test_old_date_redirect_uses_cache_for_second_query() {
```
Comment (spacedmonkey) [inline tests/phpunit/tests/rewrite/oldDateRedirect.php:87]: A covers and ticket is needed here.

File: tests/phpunit/tests/rewrite/oldDateRedirect.php [hunk 8]
```diff
@@ -86,6 +118,36 @@ public function test_old_date_slug_redirect() {
 		$this->assertSame( $permalink, $this->old_date_redirect_url );
 	}
 
+	public function test_old_date_slug_redirect_uses_cache_for_second_query() {
```
Comment (spacedmonkey) [inline tests/phpunit/tests/rewrite/oldDateRedirect.php:143]: A covers and ticket is needed here.

File: src/wp-includes/query.php [hunk 9]
```diff
@@ -1117,15 +1105,40 @@ function _find_post_by_old_slug( $post_type ) {
 		$query .= $wpdb->prepare( ' AND DAYOFMONTH(post_date) = %d', get_query_var( 'day' ) );
 	}
 
+	$key          = md5( $query );
+	$cache_key = "_find_post_by_old_slug:$key";
```
Comment (spacedmonkey) [inline src/wp-includes/query.php]:
```suggestion
        $last_changed = wp_cache_get_last_changed( 'posts' );

	$hash      = md5( $query );
	$cache_key = "_find_post_by_old_slug:$hash:$last_changed";
```

This needs to be like this to ensure that caches are invalidated.

File: src/wp-includes/query.php [hunk 10]
```diff
@@ -1148,17 +1161,42 @@ function _find_post_by_old_date( $post_type ) {
 		$date_query .= $wpdb->prepare( ' AND DAYOFMONTH(pm_date.meta_value) = %d', get_query_var( 'day' ) );
 	}
 
-	$id = 0;
 	if ( $date_query ) {
-		$id = (int) $wpdb->get_var( $wpdb->prepare( "SELECT post_id FROM $wpdb->postmeta AS pm_date, $wpdb->posts WHERE ID = post_id AND post_type = %s AND meta_key = '_wp_old_date' AND post_name = %s" . $date_query, $post_type, get_query_var( 'name' ) ) );
+		$key          = md5( $date_query );
+		$cache_key    = "_find_post_by_old_date:$key";
+		$cache        = wp_cache_get( $cache_key, 'posts' );
```
Comment (spacedmonkey) [inline src/wp-includes/query.php]:
```suggestion
                $last_changed = wp_cache_get_last_changed( 'posts' );
                $hash         = md5( $date_query );
	        $cache_key = "_find_post_by_old_date:$hash:$last_changed";
		$cache        = wp_cache_get( $cache_key, 'posts' );
```

File: src/wp-includes/query.php [hunk 11]
```diff
@@ -1148,17 +1161,42 @@ function _find_post_by_old_date( $post_type ) {
...
+			 */
+			$id = apply_filters( 'old_slug_redirect_post_id', $id );
+
+			if ( ! $id ) {
+				return;
+			}
+
+			$link = get_permalink( $id );
+
+			wp_cache_set( $cache_key, $link, 'posts' );
```
Comment (spacedmonkey) [inline src/wp-includes/query.php]:
```suggestion
			wp_cache_set( $cache_key, $id, 'posts' );
```

Again, we should only be caching the ids here not the link.

File: src/wp-includes/query.php [hunk 12]
```diff
@@ -1117,15 +1105,40 @@ function _find_post_by_old_slug( $post_type ) {
...
+	wp_cache_set( $cache_key, $link, 'posts' );
+
+	return $link;
 }
 
 /**
  * Find the post ID for redirecting an old date.
  *
  * @since 4.9.3
+ * @since 6.0.0 Added Caching Ticket #36723.
```
Comment (spacedmonkey) [inline src/wp-includes/query.php]:
```suggestion
 * @since 6.0.0 Results are are cached.
```

File: src/wp-includes/query.php [hunk 13]
```diff
@@ -1117,15 +1105,40 @@ function _find_post_by_old_slug( $post_type ) {
...
 	$id = (int) $wpdb->get_var( $query );
 
-	return $id;
+	/**
+	 * Filters the old slug redirect post ID.
+	 *
+	 * @since 4.9.3
+	 *
+	 * @param int $id The redirect post ID.
+	 */
```
Comment (spacedmonkey) [inline src/wp-includes/query.php:1123]: Filter docs do not need to be repeated. You are just reference the original docs.

**Comment by spacedmonkey** [comment]
There are php lint errors in this PR. Can you run `composer lint` / `composer format` locally to fix these.

File: src/wp-includes/query.php [hunk 14]
```diff
@@ -1043,27 +1043,16 @@ function wp_old_slug_redirect() {
 			return;
 		}
 
-		$id = _find_post_by_old_slug( $post_type );
+		$link = _find_post_by_old_slug( $post_type );
 
-		if ( ! $id ) {
-			$id = _find_post_by_old_date( $post_type );
+		if ( ! $link ) {
+			$link = _find_post_by_old_date( $post_type );
```
Comment (spacedmonkey) [inline src/wp-includes/query.php:1049]: These changes need to be reverted.

File: src/wp-includes/query.php [hunk 15]
```diff
@@ -1043,27 +1043,16 @@ function wp_old_slug_redirect() {
...
-		 * @param int $id The redirect post ID.
-		 */
-		$id = apply_filters( 'old_slug_redirect_post_id', $id );
-
-		if ( ! $id ) {
+		if ( ! $link ) {
 			return;
 		}
 
-		$link = get_permalink( $id );
```
Comment (spacedmonkey) [inline src/wp-includes/query.php:1065]: Revert this

File: src/wp-includes/query.php [hunk 16]
```diff
@@ -1079,10 +1068,6 @@ function wp_old_slug_redirect() {
 		 */
 		$link = apply_filters( 'old_slug_redirect_url', $link );
 
-		if ( ! $link ) {
-			return;
-		}
```
Comment (spacedmonkey) [inline src/wp-includes/query.php:1084]: Revert this.

**Comment by spacedmonkey** [comment]
I have own version of this patch at - https://github.com/WordPress/wordpress-develop/pull/2448

This includes unit tests.

**Comment by spacedmonkey** [comment]
Committed.

**Pull request closed without merging**